feature: User CRUD

LeonelCeballos18 opened the pull request:
## What kind of PR is this?
- [x] Feature

## Description
-add `User CRUD` in `user service`
-add `get, create, update dto` for user CRUD
-add `dto's` in index dto's
-add `filters` for `get user`

## Ticket id
https://barbershop-app.atlassian.net/jira/software/projects/SCRUM/boards/1?selectedIssue=SCRUM-25

## Useful resources
<img width="1001" alt="Screenshot 2025-04-17 at 17 13 58" src="https://github.com/user-attachments/assets/fecbc17b-aa38-4f5e-aedb-fbd2e6b4888e" />

LeonelCeballos18 committed:
feature: User CRUD

-add `User CRUD` in `user service`
-add `get, create, update dto` for user CRUD
-add `dto's` in index dto's
-add `filters` for `get user`

dsalazar357 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,86 @@ export class UserService {
+
+    const userResponse = users.map((user) => {
+      return {
+        ...user,
```
Actually you don't even need a `.map()` you can straight up write this
`const userResponse = users`

dsalazar357 replied:
Lookin at it, might not be needed, but you do you.

LeonelCeballos18 replied:
actually, that was wrong, I was not using the dto to omit the password, but I just solved

dsalazar357 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,86 @@ export class UserService {
+      };
+    });
+
+    if (param) {
```
You can do this to check for params earlier
Plus you now don't quite need userResponse, you can flatly just use `users`
```typescript
if (param && userFilters[param]) {
  return userFilters[param](users);
} else if (param) {
  throw new BadRequestException(`Parámetro de filtro inválido: ${param}`);
}
```

dsalazar357 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,86 @@ export class UserService {
+    return userResponse;
+  }
+
+  async create(crateDto: UserCreateDto): Promise<User> {
```
Correct typo here `crateDto` -> `createDto`

dsalazar357 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,86 @@ export class UserService {
+    });
+
+    if (existingUser) {
+      throw new Error('Usuario existente, usa difentes credenciales para crear uno.');
```
Correct typo `difentes` -> `diferentes`

dsalazar357 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,86 @@ export class UserService {
+      throw new Error('Usuario existente, usa difentes credenciales para crear uno.');
+    }
+    const user = this.userRepository.create(crateDto);
+    const hashedPassword = await bcryptjs.hash(user.password, 10);
```
Do this directly from the dto
```typescript
const user = this.userRepository.create(createDto);
const hashedPassword = await bcryptjs.hash(createDto.password, 10);
user.password = hashedPassword;
```

dsalazar357 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,86 @@ export class UserService {
+
+  async update(id: number, dto: UserUpdateDto): Promise<User> {
+    const user = await this.searchFor(id);
+    console.log(user);
```
Is this needed for anything? 👀

dsalazar357 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,86 @@ export class UserService {
+    try {
+      return this.userRepository.save(user);
+    } catch (error) {
+      throw new BadRequestException('Erro al actualizar el usuario');
```
correct typo `erro` -> `error`

dsalazar357 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,86 @@ export class UserService {
+    if (!user) {
+      throw new BadRequestException('Usuario no encontrado');
+    }
+    if (user) {
```
By the time the code reaches the second if (user), you've already confirmed that user exists—so there's no need to check it again.

you can do this:
```typescript
const user = await this.searchFor(id);

if (!user) {
  throw new BadRequestException('Usuario no encontrado');
}

try {
  await this.userRepository.remove(user);
} catch (error) {
  throw new InternalServerErrorException('Error al eliminar el usuario');
}
```

dsalazar357 replied:
If searchFor(id) already throws when a user is not found, then even the if (!user) check might be unnecessary:

```typescript
// if searchFor throws, you don't need this check
const user = await this.searchFor(id);

try {
  await this.userRepository.remove(user);
} catch (error) {
  throw new InternalServerErrorException('Error al eliminar el usuario');
}
```

But you'd need to implement searchFor that way first.

LeonelCeballos18 committed:
chore: Solving comments

-use `get dto` correctly to omit user password
-add `user.spec` with test for all methods
-add `id` field in `get response user dto`

cmagallon1 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,93 @@ export class UserService {
+      relations: { role: true, appointments: true },
+    });
+
+    const userResponse = users.map((user) => {
```
is it possible to do something like?:

```suggestion
const userResponse = users.map(user => new UserResponseDto(
  user.id,
  user.name,
  user.phone,
  user.email,
  user.role,
  user.appointments
));
```

cmagallon1 replied:
or: 

```suggestion
const userResponse = users.map(user => ({
  ...new UserResponseDto(),
  id: user.id,
  name: user.name,
  phone: user.phone,
  email: user.email,
  role: user.role,
  appointments: user.appointments,
}));
```

LeonelCeballos18 replied:
The first is not possible because the `UserResponseDto` class does not define a constructor that accepts arguments and would have to explicitly define a constructor in my dto.

The second one is way better, but how I did it is how it is usually done in nestjs, so the second one is not wrong, but is not the best solution as well

cmagallon1 replied:
got it, thanks for clarifying!

cmagallon1 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,93 @@ export class UserService {
+      return userResponse;
+    });
+
+    if (param && userFilters[param]) {
```
💭 I think you should validate if the param is valid before querying the database

LeonelCeballos18 replied:
Yes, you're right, I'm working on it

cmagallon1 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,93 @@ export class UserService {
+    });
+
+    if (existingUser) {
+      throw new Error('Usuario existente, usa diferentes credenciales para crear uno.');
```
errors should be in english or spanish?

LeonelCeballos18 replied:
In spanish, because fronted uses these errors to show it to users

cmagallon1 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,93 @@ export class UserService {
+      throw new Error('Usuario existente, usa diferentes credenciales para crear uno.');
+    }
+    const user = this.userRepository.create(createDto);
+    const hashedPassword = await bcryptjs.hash(createDto.password, 10);
```
why are we using the asyncronous version of the hashing method instead of encoding the password syncronously?: `hashSync`, if you use the syncronous method you can do something like:

```suggestion
    try {
    const user = this.userRepository.create({ ...createDto, password: bcryptjs.hashSync(createDto.password, 10) });
    return await this.userRepository.save(user);
    }
```

LeonelCeballos18 committed:
chore: Solving comments

-add password encoding synchronous
-add `AuthGuard` in user routes
-add validation for `filters` in `get` method

cmagallon1 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,94 @@ export class UserService {
+      return userResponse;
+    });
+
+    if (param) {
```
👀 

```suggestion
    if (!param) { return userResponse }
     
    const filterKey = param.toUpperCase();
    if (!Object.keys(userFilters).includes(filterKey)) {
      throw new BadRequestException(`Parametro Inválido: ${param}`);
    }
    return userFilters[filterKey](userResponse);
```

cmagallon1 reviewed (approved):
just a small comment, other than that LGTM

cesarj45 commented on src/dtos/user.get.response.dto.ts:
```diff
@@ -0,0 +1,36 @@
+
+export default class UserResponseDto {
+  @IsNumber()
+  @MaxLength(50)
```
@maxLength(50) is not compatible with @isNumber() because it is used for strings, not numbers; also, we should remove the password from a response DTO and I recommend using a special decorator for this type of DTO like @Exclude() and @Expose()
```suggestion
  @Exclude()
export default class UserResponseDto {
  @Expose()
  id: number;

  @Expose()
  name: string;

  @Expose()
  phone: string;

  @Expose()
  email: string;

  @Expose()
  role: Role;

  @Expose()
  @Type(() => AppointmentDto)
  appointments: AppointmentDto[];
}
```

LeonelCeballos18 committed:
Solving comments

cesarj45 commented on src/services/user.service.ts:
```diff
@@ -10,8 +17,94 @@ export class UserService {
+    });
+
+    if (!user) {
+      throw new InternalServerErrorException(`Usuario no encontrado`);
```
```
if (!user) {
    throw new NotFoundException(`Usuario no encontrado`);
  }
```
